<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="GL-2.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Faculty Notebook Week 1 <br> (Day 4) </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## Table of Content

1. **[Import Libraries](#lib)**
2. **[Test of Hypothesis](#defn)**
    - 2.1 - **[Types of Hypothesis](#types)**
3. **[Types of Test](#test_type)**
    - 3.1 - **[Two Tailed Test](#2tailed)**
    - 3.2 - **[One Tailed Test](#1tailed)**
4. **[Hypothesis Tests with Z Statistic](#eg)**
5. **[Error](#error)**

<a id="lib"></a>
# 1. Import Libraries

**Let us import the required libraries.**

In [11]:
# import 'pandas' 
import pandas as pd 

# import 'numpy' 
import numpy as np

# import subpackage of matplotlib
import matplotlib.pyplot as plt

# import 'seaborn'
import seaborn as sns

# to suppress warnings 
from warnings import filterwarnings
filterwarnings('ignore')

# import 'random' to generate random sample
import random

# import statistics to perform statistical computation  
import statistics

# import 'stats' package from scipy library
from scipy import stats

# import a library to perform Z-test
from statsmodels.stats import weightstats as stests

In [12]:
# set the plot size using 'rcParams'
# once the plot size is set using 'rcParams', it sets the size of all the forthcoming plots in the file
# pass width and height in inches to 'figure.figsize' 
plt.rcParams['figure.figsize'] = [15,8]

<a id="defn"></a>
# 2. Test of Hypothesis

Hypothesis testing is the process of evaluating the validity of claims made about the population using the sample data obtained from the population. A statistical test is a rule used to decide whether to reject or retain the claim.

**Examples of hypothesis:**

        1. One can get 'A' grade if the attendance in the class is more than 75%.
        2. A probiotic drink can improve the immunity of a person. 

<a id="types"></a>
## 2.1 Types of Hypothesis

`Null Hypothesis`: The null hypothesis is the statistical hypothesis suggesting 'no difference' between the population parameter and a specific value
. It is denoted as H<sub>0</sub>.

`Alternative Hypothesis`: It is the hypothesis that is tested against the null hypothesis and states the existence of a difference between the parameter and a specific value. It is denoted by H<sub>a</sub> or H<sub>1</sub>.

The claim is usually the alternative hypothesis H<sub>1</sub> also known as research hypothesis. To test the claim we need to gather evidence (data ) and find the likelihood of the data under the assumption H<sub>0</sub> being true.

#### Example:

A company that produces tennis balls claimed that the diameter of a tennis ball is at least 2.625 inches on average. On the other hand, a professional tennis coach claimed that the diameter of a ball is less than what the company has claimed. To test the claim of the coach, a statistical test can be performed considering the hypothesis:

                    Null Hypothesis: Average diameter ≥ 2.625
                    Alternative Hypothesis: Average diameter < 2.625

<a id="test_type"></a>
# 3. Types of Test

The hypothesis test is used to validate the claims about the population. The types of tests are based on the nature of the alternative hypothesis. 

<a id="2tailed"></a>
## 3.1 Two Tailed Test

Two tailed test considers the value of the population parameter is less than or greater than (i.e. not equal) a specific value. <br>
If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu = \mu_{0}$. 

The alternative hypothesis for the two tailed test is given as: $H_{1}: \mu \neq \mu_{0}$

#### Example:

A company that produces tennis balls claimed that the diameter of a tennis ball is 2.625 inches on average. To test the company's claim, a statistical test can be performed considering the hypothesis:

                    Null Hypothesis: Average diameter = 2.625
                    Alternative Hypothesis: Average diameter ≠ 2.625

<a id="1tailed"></a>
## 3.2 One Tailed Test

One tailed test considers the value of the population parameter is less than or greater than (but not both) a specific value. <br>
If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu \leq \mu_{0}$ and the alternative hypothesis is $H_{1}: \mu > \mu_{0}$, the one tailed test is also known as a `right-tailed test`.

If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu \geq \mu_{0}$ and the alternative hypothesis is $H_{1}: \mu < \mu_{0}$, the one tailed test is also known as a `left-tailed test`.


### Example:

**1.** The company's annual quality report of machines states that a lathe machine works efficiently at most for 8 months on average after the servicing. The production manager claims that after the special tuxan servicing, the machine works efficiently for more than 8 months. To test the claim of production manager consider the hypothesis:

                    Null Hypothesis: Machine efficiency ≤ 8 months
                    Alternative Hypothesis: Machine efficiency > 8 months

This is the example of a **right-tailed test**. 

**2.** A railway authority claims that all the trains on the Chicago-Seattle route run with a speed of at least 54 mph on average. A customer forum declares that there are various records from passengers claiming that the speed of the train is less than what railway has claimed. In this scenario, a statistical test can be performed to test the claim of customer forum considering the hypothesis:

                    Null Hypothesis: Speed ≥ 56 mph
                    Alternative Hypothesis: Speed < 56 mph

This is the example of a **left-tailed test**. 

<a id="eg"></a>
# 4. Hypothesis Tests with Z Statistic

Let us perform one sample Z test for the population mean. We compare the population mean with a specific value. The sample is assumed to be taken from a population following a normal distribution.

To check the normality of the data, a test for normality is used. The `Shapiro-Wilk Test` is one of the methods used to check the normality. The hypothesis of the test is given as:
<p style='text-indent:25em'> <strong> H<sub>0</sub>:  The data is normally distributed </strong> </p>
<p style='text-indent:25em'> <strong> H<sub>1</sub>:  The data is not normally distributed </strong> </p>

The `shapiro()` from scipy library performs a Shapiro-Wilk normality test. 

The null and alternative hypothesis of Z-test is given as:
<p style='text-indent:25em'> <strong> $H_{0}: \mu = \mu_{0}$ or $\mu \geq \mu_{0}$ or $\mu \leq \mu_{0}$</strong></p>
<p style='text-indent:25em'> <strong> $H_{1}: \mu \neq \mu_{0}$ or $\mu < \mu_{0}$ or $\mu > \mu_{0}$</strong></p>

Consider a normal population with standard deviation $\sigma$. Let us take a sample of size n, such that (n > 30). 
The test statistic for one sample Z-test is given as:
<p style='text-indent:25em'> <strong> $Z = \frac{\overline{X} -  \mu}{\frac{\sigma}{\sqrt(n)}}$</strong></p>

Where, <br>
$\overline{X}$: Sample mean<br>
$\mu$: Specified mean<br>
$\sigma$: Population standard deviation<br>
$n$: Sample size

Under $H_{0}$ the test statistic follows a standard normal distribution.

If $\sigma$ is unknown, use the sample standard deviation (s) instead of $\sigma$ to calculate the test statistic.

### Example:

#### 1. The Average IQ of Adult Population is 100 with a Standard Deviation of 15. A Researcher believes that it has changed. He decided to test the IQ of 75 random adults. The Average IQ of the sample came out to be 105. Is there enough evidence to believe that it has changed.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu =  100 $<br>
H<sub>1</sub>: $\mu != 100 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

In [13]:
# Calculate the critical value
# using isf: it stand for 'Inverse Survival Function'
# using ppf: it stand for 'Percentage point Function'
np.abs(stats.norm.ppf(0.025))

1.9599639845400545

In [14]:
stats.norm.sf(0.025)

0.49002748180476197

In [15]:
mew=100
x_bar=105
sigma=15
alpha_by2=0.025
cl=0.95
n=75

In [16]:
teststats= (x_bar-mew)/(sigma/np.sqrt(n))
print('Tesrt Statistic',teststats)

Tesrt Statistic 2.886751345948129


In [17]:
# Confidence
stats.norm.interval(confidence=cl,loc=x_bar,scale=sigma/np.sqrt(n))

(101.60524279777148, 108.39475720222852)

In [18]:
### Make a function
def humaraztest(x_bar, mew, sigma, n, alpha, test_type):
    if test_type==2:
        cv=1-(alpha)/2
        critical=np.abs(stats.norm.isf(cv))
    else: 
        cv=1-(alpha)
        critical=np.abs(stats.norm.isf(cv))
    print('z Critical',critical)
    
    # Test Statistic
    se=sigma/np.sqrt(n)
    teststats=np.abs((x_bar-mew)/se)
    
    print('Test_statistic',teststats)
    # Compare the result
    
    if teststats>critical:
        print('Reject the Ho')
    else:
        print('Fail to Reject the Ho')
        
    # Confidence Interval
    print(stats.norm.interval(cv,loc=x_bar,scale=se))


In [19]:
humaraztest(105,100,15,75,0.05,2)

z Critical 1.959963984540054
Test_statistic 2.886751345948129
Reject the Ho
(101.11777659556478, 108.88222340443524)


In [20]:
# calculate the z-value for 95% of confidence level
# use 'stats.norm.isf()' to find the z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.05
# use 'round()' to round-off the value to 2 digits

In [21]:
# Find the Critical Value
#import scipy.stats as stats

alpha = 0.05
ci = alpha/2

# Since the Population Standard Deviation is given, we will use Z Critical
# isf - Isotonic Survival Function
# ppf - Percent Point Function

z_crit = stats.norm.isf(ci) # Z Critical Value/Zscore

i.e. if z is greater than 1.96 then we reject the null hypothesis.

In [22]:
# Apply the Statistical Test

popmean = 100
samplemean = 105
popsd = 15
n = 75

teststats = (samplemean - popmean)/(popsd/np.sqrt(n))
print("TestStats: ", teststats)

TestStats:  2.886751345948129


In [23]:
# calculate the 95% confidence interval for the population mean
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (standard deviation of sample / n^(1/2)) to the parameter, 'scale'
# use 'stdev()' to calculate sample standard deviation
# Find the confidence interval

stats.norm.interval(0.95, loc = samplemean, scale = popsd/np.sqrt(n))

(101.60524279777148, 108.39475720222852)

In [24]:
# Make a Function
def ztest(samplemean, popmean, popsd, n, alpha, tail):
    # Z Critical
    if tail==2:
        ci = alpha/2
        z_crit = np.abs(stats.norm.isf(ci))
        print("Z Critical: ", z_crit)
    else:
        ci = alpha
        z_crit = np.abs(stats.norm.isf(ci))
        print("Z Critical: ", z_crit)
    
    # Test Stats
    se = popsd/np.sqrt(n)
    test_statistic = np.abs((samplemean - popmean)/se)
    print("Test Statistic: ", test_statistic)
    
    # Confidence Interval
    print("Confidence Interval at", 1 - alpha, "is ", \
          stats.norm.interval(1 - alpha,loc = samplemean,scale  =se))
    
    # PValue
    print("PValue: ", (1 - stats.norm.cdf(test_statistic))*2)
    
    # Inference:
    if test_statistic>z_crit:
        print("Reject Ho")
    else:
        print("Fail to Reject Ho")

In [25]:
ztest(105, 100, 15, 75, 0.05,2)

Z Critical:  1.9599639845400545
Test Statistic:  2.886751345948129
Confidence Interval at 0.95 is  (101.60524279777148, 108.39475720222852)
PValue:  0.0038924171227785465
Reject Ho


Here the z test statistic is greater than 1.96, thus, we reject the null hypothesis. Further, the Confidence Interval Range also supports the Evidence that the Population Mean is not 100 anymore.

**Note: Please note that when null hypothesis is not rejected it cannot be accepted as true.There is merely not enough evidence to say that it is false. This may seem confusing but the situation is analogous to the court of law, where the verdict is either guilty or not guilty based on the evidence. If the person is judged not guilty then it does not mean the person is proved innocent but merely that there is not enough evidence to reach the guilty verdict.**

#### 2. A car manufacturing company claims that the mileage of their new car is 25 kmph with a standard deviation of 2.5 kmph. A random sample of 45 cars was drawn and recorded their mileage as per the standard procedure. From the sample, the mean mileage was seen to be 24 kmph. Is this evidence to claim that the mean mileage is different from 25kmph? (assume the normality of the data) Use α = 0.01.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu = 25 $<br>
H<sub>1</sub>: $\mu ≠ 25 $

Here ⍺ = 0.01, for a two-tailed test calculate the critical z-value.

In [26]:
# popmean = 25, popsd = 2.5, n = 45, samplemean = 24

humaraztest(24,25,2.5,45,0.01,2)

z Critical 2.5758293035489004
Test_statistic 2.6832815729997477
Reject the Ho
(22.953880279807642, 25.046119720192358)


#### Conclusion:

* We reject the Ho basis the output however, we see that the CI Range Upper Limit is 25 which suggests that it is resulting into an error

* Thus, we need to increase the sample size for better results.

In [27]:
# popmean = 25, popsd = 2.5, n = 45, samplemean = 24

ztest(24, 25, 2.5, 45, 0.01,2)

Z Critical:  2.575829303548901
Test Statistic:  2.6832815729997477
Confidence Interval at 0.99 is  (23.040045096471452, 24.959954903528548)
PValue:  0.007290358091535554
Reject Ho


In [28]:
# Calculate the Test Statistic


In [29]:
# calculate the 99% confidence interval for the population mean


#### 3. The average calories in a slice bread of the brand 'Alphas' are 82 with a standard deviation of 15. An experiment is conducted to test the claim of the dietitians that the calories in a slice of bread are not as per the manufacturer's specification. A sample of 40 slices of bread is taken and the mean calories recorded are 95. Test the claim of dietitians with ⍺ value (significance level) as 0.05. (assume the normality of the data).

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu = 82 $<br>
H<sub>1</sub>: $\mu ≠ 82 $

Here ⍺ = 0.05, for a two-tailed test calculate the critical z-value.

In [30]:
humaraztest(95,82,15,40,0.05,2)

z Critical 1.959963984540054
Test_statistic 5.4812812776251905
Reject the Ho
(89.68404667029081, 100.3159533297092)


In [31]:
# calculate the z-value for 95% of confidence level
# use 'stats.norm.isf()' to find the z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha/2' to the parameter 'q', here alpha = 0.05
# use 'round()' to round-off the value to 2 digits


In [32]:
ztest(95, 82, 15, 40, 0.05, 2)

Z Critical:  1.9599639845400545
Test Statistic:  5.4812812776251905
Confidence Interval at 0.95 is  (90.35153725771579, 99.64846274228421)
PValue:  4.222565253009236e-08
Reject Ho


In [33]:
# calculate the 95% confidence interval for the population mean


#### 5. An e-commerce company claims that the mean delivery time of food items on its website in NYC is 60 minutes with a standard deviation of 30 minutes. A random sample of 45 customers ordered from the website, and the average time for delivery was found to be 75 minutes. Is this enough evidence to claim that the average time to get items delivered is more than 60 minutes. (assume the normality of the data). Test the with α = 0.05.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \leq  60 $<br>
H<sub>1</sub>: $\mu > 60 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

In [34]:
humaraztest(75,60,30,45,0.05,1)

z Critical 1.6448536269514722
Test_statistic 3.3541019662496843
Reject the Ho
(66.23477459423418, 83.76522540576582)


In [35]:
# calculate the z-value for 95% of confidence level
# use 'stats.norm.isf()' to find the z-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.05
# use 'round()' to round-off the value to 2 digits


In [39]:
ztest(75, 60, 30, 45, 0.05,1)

Z Critical:  1.6448536269514729
Test Statistic:  3.3541019662496843
Confidence Interval at 0.95 is  (66.23477459423418, 83.76522540576582)
PValue:  0.0007962301575907826
Reject Ho


# One Sample T Test

<a id="t"></a>
## 1. Small Sample Test

If the sample size is small (usually, n < 30) then we use the `t-test`. These tests are also known as `exact tests`.

Let us perform a one sample t-test for the population mean. We compare the population mean with a specific value. 

The null and alternative hypothesis is given as:

<p style='text-indent:25em'> <strong> $H_{0}: \mu = \mu_{0}$ or $\mu \geq \mu_{0}$ or $\mu \leq \mu_{0}$</strong></p>
<p style='text-indent:25em'> <strong> $H_{1}: \mu \neq \mu_{0}$ or $\mu < \mu_{0}$ or $\mu > \mu_{0}$</strong></p>

The test statistic is given as:
<p style='text-indent:25em'> <strong> $t = \frac{\overline{X} -  \mu_{0}}{\frac{s}{\sqrt(n)}}$</strong></p>

Where, <br>
$\overline{X}$: Sample mean<br>
$s$: Sample standard deviation<br>
$n$: Sample size
 
Under $H_{0}$ the test statistic follows a t-distribution with n-1 degrees of freedom.

#### 6. Experience marketing Services reported that the typical American spend a mean of 144 minutes(2.4 hours) per day accessing the internet via mobile device. In order to test the validity of the statement, you select a sample of 30 friends and family. The result for the time spent per day accessing the Internet via mobile device (in minutes) are stored. Is there enough evidence that Pop Mean Time spent Per Day accessing the internet via mobile device is different from 144 minutes.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu=144 $<br>
H<sub>1</sub>: $\mu!=144 $

In [41]:
# Time in Mins

mins = [72,144,48,72,36,360,44,30,432,24,288,144,144,240,432,144,
        144,144,576,216,72,72,144,288,144,36,288,48,288,144]

# Calculate the Sample Standard Deviation

# Calculate the Standard Error....
np.std(mins)

137.48646802099796

In [42]:
humaraztest(144,175.26,137.48,30,0.05,2)

z Critical 1.959963984540054
Test_statistic 1.2454034876063056
Fail to Reject the Ho
(87.74012120393246, 200.2598787960676)


In [43]:
# Find the Critical Value

np.abs(stats.t.isf(0.95, len(mins)-1))

1.6991270265334972

In [44]:
# Calculate the Test Statistic

stats.ttest_1samp(mins, 144)

TtestResult(statistic=1.2246743653638934, pvalue=0.23055326882983754, df=29)

In [45]:
# Confidence

stats.t.interval(0.95, loc = np.mean(mins), 
                scale = np.std(mins, ddof = 1)/np.sqrt(len(mins)), 
                df = len(mins)-1)

(123.05073459043008, 227.4825987429033)

### Inference: 

* Since, the P Value comes out to be 0.23 which is much higher than  alpha = 0.05. Hence, we Fail to Reject the HO meaning that the Average Time Spent by an American on the Internet Using Mobile is really 144  Mins.

* At the same time, the confidence interval also reflects the Hypothesized Mean which also validates the above results.

<a id="error"></a>
# 5. Errors in Hypothesis Testing

Under the hytpothesis testing framework there can be two decisions; either we 'reject' the null hypotyhesis H<sub>0</sub> or 'not reject' the null hypothesis H<sub>0</sub>. Errors can still occur in a hypothesis testing scenario because the decision to reject or not reject the null hypothesis is made on the basis of data taken from a sample. Two types of errors are `Type I` and `Type II` error.

### Type I Error

This kind of error occurs when we reject the null hypothesis even if it is true. It is equivalent to a `false positive` conclusion. The maximum probability of committing a type I error is given by the value of $\alpha$, level of significance. So when $\alpha$ = 0.05 there is a 5% chance of rejecting a true null hypothesis.

### Type II Error

This kind of error occurs when we fail to reject the null hypothesis even if it is wrong. It is equivalent to a `false negative` conclusion. The probability of type II error is given by the value of $\beta$. The value of $\beta$ cannot be easily calculated. But $\alpha$ and $\beta$ are thus related that decreasing one increases the other.